File: Olympiad/COCI/official/2009/regional_solutions/cijevi.pas
program cijevi;

const MAXR = 25;
const MAXS = 25;

var
   R, S, i, j : longint;
   a : array[1..MAXR, 1..MAXS+1] of char;

function lijevo( i, j : longint ) : boolean;
begin
   if (j > 1) and ((a[i,j-1]='+') or (a[i,j-1]='-') or (a[i,j-1]='1') or (a[i,j-1]='2')) then lijevo := true else lijevo := false;
end;
function desno( i, j : longint ) : boolean;
begin
   if (j < S) and ((a[i,j+1]='+') or (a[i,j+1]='-') or (a[i,j+1]='3') or (a[i,j+1]='4')) then desno := true else desno := false;
end;
function gore( i, j : longint ) : boolean;
begin
   if (i > 1) and ((a[i-1,j]='+') or (a[i-1,j]='|') or (a[i-1,j]='1') or (a[i-1,j]='4')) then gore := true else gore := false;
end;
function dolje( i, j : longint ) : boolean;
begin
   if (i < R) and ((a[i+1,j]='+') or (a[i+1,j]='|') or (a[i+1,j]='2') or (a[i+1,j]='3')) then dolje := true else dolje := false;
end;

begin
   readln( R, S );
   for i := 1 to R do readln( a[i] );
   
   for i := 1 to R do begin
      for j := 1 to S do begin
         if a[i,j] <> '.' then continue;

         if lijevo(i,j) and desno(i,j) and gore(i,j) and dolje(i,j) then writeln( i, ' ', j, ' +' )
         else if lijevo(i,j) and desno(i,j) then writeln( i, ' ', j, ' -' )
         else if   gore(i,j) and dolje(i,j) then writeln( i, ' ', j, ' |' )
         else if  desno(i,j) and dolje(i,j) then writeln( i, ' ', j, ' 1' )
         else if  desno(i,j) and  gore(i,j) then writeln( i, ' ', j, ' 2' )
         else if lijevo(i,j) and  gore(i,j) then writeln( i, ' ', j, ' 3' )
         else if lijevo(i,j) and dolje(i,j) then writeln( i, ' ', j, ' 4' );
      end;
   end;  
end.

File: Olympiad/COCI/official/2009/regional_solutions/cijevi.cpp
#include <cstdio>

#define MAXR 25
#define MAXS 25

int R, S;
char a[MAXR][MAXS+1];
int lijevo( int r, int s ) {
   return s >  0  && (a[r][s-1]=='+' || a[r][s-1]=='-' || a[r][s-1]=='1' || a[r][s-1]=='2');
}
int desno( int r, int s ) {
   return s < S-1 && (a[r][s+1]=='+' || a[r][s+1]=='-' || a[r][s+1]=='3' || a[r][s+1]=='4');
}
int gore( int r, int s ) {
   return r >  0  && (a[r-1][s]=='+' || a[r-1][s]=='|' || a[r-1][s]=='1' || a[r-1][s]=='4');
}
int dolje( int r, int s ) {
   return r < R-1 && (a[r+1][s]=='+' || a[r+1][s]=='|' || a[r+1][s]=='2' || a[r+1][s]=='3');
}

int main( void ) {
   scanf( "%d%d", &R, &S );
   for( int r = 0; r < R; ++r ) scanf( "%s", a[r] );

   for( int r = 0; r < R; ++r ) 
      for( int s = 0; s < S; ++s ) {
         if( a[r][s] != '.' ) continue;

         if( lijevo(r,s) && desno(r,s) && gore(r,s) && dolje(r,s) ) printf( "%d %d +\n", r+1, s+1 );
         else if( lijevo(r,s) && desno(r,s) ) printf( "%d %d -\n", r+1, s+1 ); 
         else if(   gore(r,s) && dolje(r,s) ) printf( "%d %d |\n", r+1, s+1 );
         else if(  desno(r,s) && dolje(r,s) ) printf( "%d %d 1\n", r+1, s+1 );
         else if(  desno(r,s) &&  gore(r,s) ) printf( "%d %d 2\n", r+1, s+1 );
         else if( lijevo(r,s) &&  gore(r,s) ) printf( "%d %d 3\n", r+1, s+1 );
         else if( lijevo(r,s) && dolje(r,s) ) printf( "%d %d 4\n", r+1, s+1 );

      }
   return 0;
}
